libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,616 @@
+#include "can_datagram.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "crc32.h"
+#include "fsm.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+#include "watchdog.h"
+#include "fifo.h"
+
+#define PROTOCOL_VERSION_SIZE_BYTES 1
+#define CRC_SIZE_BYTES 4
+#define DGRAM_TYPE_SIZE_BYTES 1
+#define DEST_LEN_SIZE_BYTES 1
+#define MAX_DEST_NODES_SIZE_BYTES 255
+#define DATA_LEN_SIZE_BYTES 2
+#define MAX_DATA_SIZE_BYTES 2048
+#define MAX_CAN_TX 4
+
+#define DATAGRAM_RX_BUFFER_LEN 2048 
+
+#define MAX_CRC_STREAM_SIZE_BYTES \
+DGRAM_TYPE_SIZE_BYTES + DEST_LEN_SIZE_BYTES \
++ DATA_LEN_SIZE_BYTES + 2*sizeof(uint32_t)
+
+#define RX_WATCHDOG_TIMEOUT_MS 25
+#define CAN_BUFFER_SIZE 8
+
+static Fsm s_dgram_fsm;
+static CanDatagramStorage s_store;
+static WatchdogStorage s_watchdog;
+
+static uint8_t s_can_buffer[CAN_BUFFER_SIZE];
+
+static Fifo s_fifo;
+static uint8_t s_tx_buffer[DATAGRAM_RX_BUFFER_LEN];
```

Comment:
@ryandancy With the new structure, the fifo where data is and process is 2048 bytes long. It is one instance which is used for both tx/rx (I am renaming it to s_buffer), and all received/transmitted data is processed through here. 

If I make it any smaller than 2048, then it overflows when the `data` segment received is large. This also has the added bonus of allowing me to read data and destination_node buffers in one call to fifo_pop_arr()  *See below*